### DISCUSSION TITLE
When using Pyright in `Pre-commit.ci`, can it ignore third-party module checks? Something like `--ignore-missing-imports`?

### DISCUSSION THREAD

**Discussion by WSH032** [microsoft/pyright#6394] [Q&A]
When I try to use `Pyright` in `pre-commit.ci`, I'm knew that `Pyright` is installed in an isolated environment, so it cannot access third-party libraries.  
But I noticed that in the [`pre-commit hook`](https://github.com/pre-commit/mirrors-mypy#using-mypy-with-pre-commit) for `mypy`, there is a `--ignore-missing-imports` parameter to skip checking third-party libraries.

Are there **similar command-line parameters** for `Pyright`?

I knew there is `'reportMissingImports': false` configuration for Pyright currently,
but I prefer not to add it in `pyproject.toml` because `pyproject.toml` will conduct the usual `Pyright` checks in a genuine Python environment.

But, using another `pyrightconfig.json` file alongside seems a bit redundant.

![}}C~OF 1R$9MO9~LUWQBLT4](https://github.com/microsoft/pyright/assets/126865849/754d0360-d2fe-4904-85af-cc37cd9e6119)

---

The more critical issue is that only adding  `'reportMissingImports': false` isn't sufficient. 
`Pyright` still reports other problems, such as `reportGeneralTypeIssues` and `reportUntypedFunctionDecorator`.

I'm unsure if **there are still additional ignore configurations** that need to be added?

```py
from typing import Union

import httpx
from fastapi import FastAPI

app = FastAPI()


def foo(url: Union[str, httpx.URL]) -> None:
    if isinstance(url, str):
        url = httpx.URL(url)
    print(url.host)  # reportGeneralTypeIssues


@app.get("/")  # reportUntypedFunctionDecorator
def bar():
    pass

```

**Comment by erictraut** [comment]
Pyright never type checks third-party libraries, so providing a mechanism to tell it to skip checking of third-party libraries would have no effect.

Pyright will follow imports to get type information from third-party libraries if they're present. If an imported module cannot be found, all symbols imported from it will be treated as `Unknown`. This can, in turn, cause other errors in your own code to be flagged depending on the options you have selected — as you've demonstrated above. You could turn off some of the diagnostic checks and add necessarily `# type: ignore` comments to suppress the remaining issues.

Most users prefer to make their type checking output consistent between their local editing experience and their CI. I find it surprising that you're intentionally deviating between these two environments. This will result in surprises in your CI scripts that are difficult to recreate locally. That sounds to me like a really painful development experience. I sure wouldn't want to impose that pain on my team!

Is it not feasible to change your CI scripts to install dependencies for your library before it runs pyright? Presumably, you're also running test cases in your CI as well, and those also require dependencies to be installed, so this would be no different.

Alternatively, you could create type stubs for the libraries you depend on and check these stubs into a `typings` directory within your repo. This way, you would always see consistent behavior regardless of whether the libraries were installed. This is an approach I've used in past projects.

**Reply by WSH032** [reply]
Thank you very much for your quick response!

---

> Most users prefer to make their type checking output consistent between their local editing experience and their CI. I find it surprising that you're intentionally deviating between these two environments. This will result in surprises in your CI scripts that are difficult to recreate locally. That sounds to me like a really painful development experience. I sure wouldn't want to impose that pain on my team!

> Is it not feasible to change your CI scripts to install dependencies for your library before it runs pyright? Presumably, you're also running test cases in your CI as well, and those also require dependencies to be installed, so this would be no different.

This is indeed a problem, but it's not my intention. I find it a bit silly to install dependencies twice, once in the virtual environment and then again in pre-commit. Some dependencies can be quite large (like PyTorch, over 500 MB), which might not be an issue for CI running in online environments such as GitHub Actions, but it might not be suitable for local development environments (downloading PyTorch twice can be a bit frustrating) .

So, I've decided to let the local pre-commit handle some simple, dependency-free basic checks (e.g `Ruff`, `Black`, `codespell`) . For checks involving dependencies like `pytest` and others (e.g  a full Pyright check involving third-party dependencies.), I'll use GitHub Actions to manage those.

Initially, I excluded Pyright from the pre-commit because the documentation specified that Pyright needs to be within the environment to perform checks. Later, I discovered that MyPy's pre-commit hook, by default, uses `--ignore-missing-imports` to circumvent the dependency requirements. This led me to explore the feasibility of Pyright in this aspect.

**Reply by WSH032** [reply]
> Most users prefer to make their type checking output consistent between their local editing experience and their CI.

Locally, there is indeed consistent type checking with the CI (both use the same rules set in `pyproject.toml`) , but it's in a different script rather than within the `pre-commit`. So, as long as the rules for `Pyright` in the `pre-commit` are a subset of the rules in the CI (meaning the `pre-commit` rules will be more lenient, and passing the CI check requires passing the pre-commit check), it should suffice.

---

Perhaps setting a separate, basic `pyrightconfig.json` for `pre-commit` would be better? Instead of modifying rules in the `pyproject.toml` to accommodate `pre-commit` (as I did in the image at the very top.)?

```json
{
  "reportMissingImports": false,
  "reportUnknownArgumentType": false,
  "reportUnknownVariableType": false,
  "reportUnknownMemberType": false,
  "reportUnknownParameterType": false,
  "reportGeneralTypeIssues": false,
  "typeCheckingMode": "basic",
  "pythonVersion": "3.8"
}
```

**Reply by erictraut** [reply]
Are you running the pre-commit hook locally within your dev environment? If so, why would you need to install dependencies again? Haven't you (or the other devs on the team) already installed the dependencies locally? Why not just run pyright using your local Python environment.

I guess I've never used pre-commit hooks, and the desired workflow isn't clear to me. My preferred workflow is to set `python.analysis.diagnosticMode` to "workspace" in Pylance (or the Pyright VS Code extension or pyright language server, if you don't want to use Pylance). Using this approach, any type-related errors are reported in my editor. If no errors are reported, then I have high confidence that my CI scripts will likewise find no errors.

A few of my former teammates preferred to use Pycharm over VS Code. They couldn't run pyright in the editor, so they'd run `pyright --watch` in a terminal window. As they saved files in Pycharm, pyright would (re)analyze the files for type errors. Once all errors were gone, the code could be submitted for a PR.

**Reply by WSH032** [reply]
I run `pre-commit` both locally and on GitHub (on github, it will be [apps/pre-commit-ci](https://github.com/apps/pre-commit-ci)).

In my local setup, I can indeed specify an existing environment using `venvPath` or the `python.analysis` you mentioned. However, it might not be feasible to do so for `apps/pre-commit-ci`, and the dependency environment paths for other team members might differ from mine. Therefore, I can't pre-specify `venvPath` for Pyright in the `.pre-commit-config.yaml` (the configuration file for pre-commit)."

---

> Why not just run pyright using your local Python environment.

Indeed, I do run Pyright through my local Python environment (the type checking script that I mentioned earlier, consistent with CI). However, I'm hoping that pre-commit can help me prevent commits that don't pass the check when I commit to Git.

Therefore, I want to run Pyright in pre-commit (perhaps my behavior seems a bit foolish, but it satisfies my obsession with clean commits)

---

> I guess I've never used pre-commit hooks, and the desired workflow isn't clear to me. My preferred workflow is to set python.analysis.diagnosticMode to "workspace" in Pylance (or the Pyright VS Code extension or pyright language server, if you don't want to use Pylance). Using this approach, any type-related errors are reported in my editor. If no errors are reported, then I have high confidence that my CI scripts will likewise find no errors.

My workflow involves conducting basic code checks via pre-commit before a Git commit. Upon pushing to GitHub, I conduct a comprehensive check and testing using GitHub Actions.

I do use VS Code, so getting real-time feedback from `Pyright` isn't an issue for me. I just want `pre-commit` to remind me of any overlooked errors before a git commit.

**Reply by jamesowers-cohere** [reply]
In case someone else gets here, the 3rd party pyright pre-commit hook allows you to point to the venvPath:

https://github.com/RobertCraigie/pyright-python?tab=readme-ov-file#pre-commit

**Reply by WSH032** [reply]
@jamesowers-cohere The problem here is that in [pre-commit.ci](https://pre-commit.ci/), there is no actual `venv` (i.e., dependencies).

my final solution: dont use pre-commit.ci🤣

**Reply by jamesowers-cohere** [reply]
That's quite easily solved though! You just install to your venv in a prior step.

**Reply by jamesowers-cohere** [reply]
(I may be misunderstanding though - I just just pre-commit within a GitHub action workflow)

**Reply by WSH032** [reply]
> You just install to your venv in a prior step.

Can't do that in `pre-commit.ci` 🤷‍♂️

---

I just just pre-commit within a GitHub action workflow.

I do the same thing like you 🤣.
